manta-pay/src/crypto/encryption/aes.rs

```diff
@@ -0,0 +1,167 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! AES Encryption Implementation
+
+// FIXME: Don't use raw bytes as encryption/decryption key.
+
+use aes_gcm::{
+    aead::{Aead, NewAead},
+    Aes256Gcm, Nonce,
+};
+use core::convert::Infallible;
+use manta_crypto::{
+    encryption::{
+        CiphertextType, Decrypt, DecryptedPlaintextType, DecryptionKeyType, Derive, Encrypt,
+        EncryptionKeyType, HeaderType, PlaintextType, RandomnessType,
+    },
+    rand::{RngCore, Sample},
+};
+use manta_util::{
+    codec::{Decode, DecodeError, Encode, Read, Write},
+    Array,
+};
+
+/// AES-GCM Authentication Tag Size
+#[allow(clippy::cast_possible_truncation)] // NOTE: GCM Tag Size should be smaller than `2^32`.
+const TAG_SIZE: usize = (aes_gcm::C_MAX - aes_gcm::P_MAX) as usize;
+
+/// Computes the size of the ciphertext corresponding to a plaintext of the given
+/// `plaintext_size`.
+#[inline]
+pub const fn ciphertext_size(plaintext_size: usize) -> usize {
+    plaintext_size + TAG_SIZE
+}
+
+/// Fixed-Nonce AES Galois Counter Mode
```

Comment:
Shall we add a `Safety` section here to inform the user that encryption key can be used only once?

Reply:
Yes, that's a good point. Use this format:

```rust
/// Fixed-Nonce AES Galois Counter Mode
///
/// # Safety
///
/// ...
```